lib/Frontend/CompilerInvocation.cpp

```diff
@@ -222,26 +222,161 @@ class ArgsToFrontendInputsConverter {
     }
     return false;
   }
+};
+    class FrontendArgsToOptionsConverter {
+    private:
+        DiagnosticEngine &Diags;
+        const llvm::opt::ArgList &Args;
+        FrontendOptions &Opts;
+        
+        llvm::Optional<const std::vector<std::string>>
+        cachedOutputFilenamesFromCommandLineOrFilelist;
+        
+
+
+  // This is a separate function so that it shows up in stack traces.
+  LLVM_ATTRIBUTE_NOINLINE
+  static void debugFailWithAssertion() {
+    // This assertion should always fail, per the user's request, and should
+    // not be converted to llvm_unreachable.
+    assert(0 && "This is an assertion!");
+  }
+
+  // This is a separate function so that it shows up in stack traces.
+  LLVM_ATTRIBUTE_NOINLINE
+  static void debugFailWithCrash() { LLVM_BUILTIN_TRAP; }
```

Comment:
Again, these two functions should remain top-level, not members.

Reply:
Moved them. But there's something here I don't understand.